### ISSUE TITLE
Export queryResult

### ISSUE THREAD

**Issue by oliversalzburg** [vitaly-t/pg-promise#41]
[Query Result Mask](https://github.com/vitaly-t/pg-promise#query-result-mask) mentions that `queryResult` can be used to construct a bitmask to qualify expectations about a query result.

I know it's trivial to just use the integer values directly, but I think it would be nice if `queryResult` was exported from the library, so that it can be used when calling `query()`.

**Comment by vitaly-t** [comment]
But it is exported from the library, into the global namespace. What makes you think that it is not? :)

**Comment by oliversalzburg** [comment]
I'm confused. I'm not seeing it being attached to `$main` or `module.exports`. But I also just noticed the `// No 'use strict' here, to let queryResult export into the global namespace.`. Well, I have to get back to this after my trip ;D

**Comment by vitaly-t** [comment]
Yep, that's the way it works, through implicit injection into the global namespace, which a module can do only without `use strict`.

**Comment by vitaly-t** [comment]
Please follow up, and close the issue, if your question is answered.

**Comment by oliversalzburg** [comment]
Why is it exported into the global namespace anyway? Wouldn't it be much cleaner to export it from the module?

If I just use `queryResult` globally in my code, I feel like that would be confusing, because people might not be aware where it is exported from.

**Comment by vitaly-t** [comment]
I did have second thoughts about myself, it it would be a breaking change. And I haven't made breaking changes in this library since ever, all changes have been incremental.

I am considering to make a number of breaking changes for version 2.0, but not sure when I will do it yet.

**Comment by oliversalzburg** [comment]
Would you consider exporting `queryResult` in addition from the module? That would make it more obvious where the symbol is coming from when we use it in our code.

Thank you very much btw for your continued work on this library. It's much appreciated!

**Comment by vitaly-t** [comment]
There is a standard way in Node JS to make it explicit that you are accessing a global variable, and that's by using `global.`, example: `global.queryResult.many`.

**Comment by oliversalzburg** [comment]
Good point about using `global.`. Thanks!

**Comment by vitaly-t** [comment]
Are we good to close the issue? If so, please close it ;)

**Comment by oliversalzburg** [comment]
Yeah, sorry. I'd be very happy to see this changed in 2.0, but I know how to work with this in the meantime :) Cheers

**Comment by vitaly-t** [comment]
Just so, you have the complete set of methods that are representing the use of those flags, including all possible combinations: `none`, `one`, `many`, `oneOrNone`, `manyOrNone`=`any` ;)

**Comment by oliversalzburg** [comment]
I know. But I wrote an abstraction layer that has a `listBy` and `readBy` method, which both bind to a generic `selectBy` method. They only differ in what they expect `selectBy` to return; a single item or many. That's where I wanted to utilize `query()` with `queryResult`. I'm already using the methods you mentioned anywhere else.

Thanks though :)

**Comment by vitaly-t** [comment]
P.S. Tip of the day : One of the most powerful features in the library - event [extend](https://github.com/vitaly-t/pg-promise#extend), you can register whole object repositories for a high-level abstraction of business objects. If you are familiar with .NET, that's like doing a proper DAC.

**Comment by oliversalzburg** [comment]
We're already making use of `extend` ;)

**Comment by vitaly-t** [comment]
Your example is a little more complicated than the one I provided for the `extend` event, using just simple object. I would probably do it in such a way as to avoid creating a new object every time also.

To answer your question, the `db` object that represents the database is always the same, it is extended only once when it is created.

But for calls like `tx` and `task` a new context is created every time, so you would end up allocating a new instance of your repository that way. That's why a little optimization here would make sense.

**Comment by oliversalzburg** [comment]
Awesome. That's already very helpful advise. Thanks again.

**Comment by vitaly-t** [comment]
@oliversalzburg Feel free to throw in suggestions here, if you have any [version 2.0 changes](https://github.com/vitaly-t/pg-promise/wiki/2.0)

**Comment by oliversalzburg** [comment]
@vitaly-t Heh, I was just drafting a very long email to you, because I thought I had found another issue with pg-promise. But the rubber-duck-debugging while writing the email lead me to the solution.

I really wish something in our postgres stack would detect/warn about connection pool exhaustion (which was the cause for our issue). But I don't think pg-promise is the right place for this.

The draft looks really nice. One thing we're using in our code, which I could see as beneficial in pg-promise is the automatic generation of task/transaction IDs. We assign a UUID to every task/transaction, which makes it slightly easier to track them in our logs.

Other than that, I've really not found anything missing from pg-promise.

**Comment by vitaly-t** [comment]
@oliversalzburg  Thank you for your feedback! 

You know, you increase the connection pool like so:
```javascript
pgp.pg.defaults.poolSize = 20;
```
see also: [pool size](https://github.com/brianc/node-postgres/blob/master/lib/defaults.js#L27)

As for the UUID usage for the tasks,...it is a step above what the library does at the moment, which is plenty already. See my [notes for 1.9.1 release](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.9.1).

The logic that's there now: Check for the tag passed, and if it is not passed, then check for the default value associated with the task (property tag), and if that one not set, then use the function name, and if the function name doesn't exist, only then you get no tag.

Also, not to forget, the tag doesn't have to be a text string, you can pass anything. But if you want its name to be recognized by pg-monitor automatically, just add method `toString()` into your custom tag object, and it will work again ;)

**Comment by oliversalzburg** [comment]
Yeah, I figured out a way to increase the pool size shortly after I realized what my problem was (which took me hours!)

I came by that auto-tagging code today, while debugging, and was kind of surprised to see those capabilities :D I might actually take another look at our tagging mechanism again and see if I can utilize the auto-tagging.

Thanks again :)

**Comment by vitaly-t** [comment]
By the way, all that tagging mechanism was described here: [Tags](https://github.com/vitaly-t/pg-promise/wiki/workflows#tags) ;)

**Comment by vitaly-t** [comment]
I've spent a weekend contemplating a proper definition of [Data Throttling versus Load Balancing](https://github.com/vitaly-t/spex/blob/master/docs/concept/throttling.md) for my new library. I appreciate if you can tell me whether it makes good sense to you, and whether or not you agree with the definition that I wrote. Thank you!!!

**Comment by oliversalzburg** [comment]
I always interpreted *load balancing* as the process of distributing tasks among multiple, usually equally capable, resources. As in, we're balancing the load of multiple resources on the same level. So, either my understanding of the term is not correct or I'm not seeing that in your definition.

The *data throttling* definition sounds spot on from my point of view.

**Comment by vitaly-t** [comment]
Thank you! It is good having a third-person perspective. I will try to improve on the definitions.

**Comment by vitaly-t** [comment]
@oliversalzburg Interesting test: https://github.com/vitaly-t/spex/blob/master/docs/concept/sequencing.md

I can't fathom how a sequence built on promises manages to outperform one that's calculated directly, but it is what's happening. Any idea? :)

**Comment by oliversalzburg** [comment]
@vitaly-t From a brief look, I would expect that the promise case allows for superior optimization for some reason. Maybe V8 can guess the final size of `result` in the promise case and saves reallocs. But that's just a wild guess.

**Comment by vitaly-t** [comment]
@oliversalzburg yeah, it got me off guard, I can't understand this myself: http://stackoverflow.com/questions/32851905/promise-outperforms-direct-calculation

Not that I'm not happy about it, more like the other way round, I just want to understand... :)

**Comment by oliversalzburg** [comment]
Join us in the StackOverflow JavaScript chat room if you don't get any good answers ;D

**Comment by vitaly-t** [comment]
Cheers! :)

**Comment by vitaly-t** [comment]
@oliversalzburg Just to poke you about the update - [version 2.0](https://github.com/vitaly-t/pg-promise/releases/tag/v.2.0.8) of the library has been released ;)

Looking forward to hearing how you get along with the new version, if you decide to upgrade ;)

**Comment by oliversalzburg** [comment]
@vitaly-t Oh, sweet. I'll check it out ASAP :)

**Comment by vitaly-t** [comment]
@oliversalzburg From what I can read in [reduce documentation](https://github.com/petkaantonov/bluebird/blob/master/API.md#reducefunction-reducer--dynamic-initialvalue---promise), it is a more specific function, while [sequence](https://github.com/vitaly-t/spex/blob/master/docs/code/sequence.md) is more generic.

**Comment by oliversalzburg** [comment]
Hah, I didn't even know about `Promise.reduce` in bluebird! :D I was thinking about code like this:

```js
return _.reduce( selection, function populateForeignKey( previous, foreignKey ) {
	var foreignType = self.foreignKeys[ foreignKey ];

	if( !foreignType ) {
		log.warn( "Invalid foreign key '" + foreignKey + "' in selection!" );
		return previous;
	}

	return previous
		.then( _.bind( self.populate, self, record, foreignKey, foreignType ) );

}, Promise.resolve( record ) );
```

I guess `Promise.reduce` would be even better, if I understood it correctly. And spex.sequence could be even more betterer ;D

So much to learn still…

**Comment by vitaly-t** [comment]
@oliversalzburg Did you get a chance to migrate to the latest 2.x version?

**Comment by oliversalzburg** [comment]
@vitaly-t We're running on 2.2.5. No issues :)

**Comment by vitaly-t** [comment]
@oliversalzburg cool! [2.3](https://github.com/vitaly-t/pg-promise/releases) is out! :)

**Comment by vitaly-t** [comment]
@oliversalzburg What do you think about [this thing](https://github.com/vitaly-t/pg-core) I just started?

It is already fully functional, by the way. I got tired with the almost non-existing support by the original author, who also banned my account, so I can no longer provide any support for users :-1:

**Comment by vitaly-t** [comment]
@oliversalzburg 

[pg-promise 3.0](https://github.com/vitaly-t/pg-promise/releases) is now available. Please let me know what you think of this whole idea...

**Comment by oliversalzburg** [comment]
I generally like the idea. I don't want to complain about node-postgres as it is very helpful in our project. However, I also noticed that the author isn't able to take care of support requests. Which is understandable, but disappointing nevertheless.

We currently have a pretty stable setup without any issues, so there isn't really any incentive to migrate to a different library.

What would really help us a lot, is a scheduling/queuing approach that solves resource conflicts caused by transactions that both want to lock the same tables. We currently only allow a single transaction at a time to work around the issue, but this is rather disappointing because it doesn't allow us to have any concurrency :( Our current system isn't smart at all and doesn't even allow concurrent read-only transactions :P

If pg-core has the same interface as node-postgres, I can definitely see us moving there as I definitely enjoy your responsiveness and the care you take of your projects :+1:

**Comment by vitaly-t** [comment]
Thank you for coming back!

About node-postgres, I believe the support provided there is awful, practically none in fact. I have been providing support there a lot to everybody, till about June this year when the author decided to ban my account, as a thanks for all my support. And when I asked him why he did it, no response ever came. This is one really bad thing here on Github when the owner is allowed to ban users without any reason.

I'm shocked about the issue with transactions you describe. I've never seen those myself, yet. I would strongly suggest to you tasks for your read-only transactions, do not over-use the transactions. Other than that, the locking needs some research to see where the problem really lies and how to approach to fix this.

If you can make time to investigate it and get some results - please log it as a separate issue, and I will be sure to check it out.

And pg-core does have the same interface. I have described the only differences right on the main page. Other than that it is all original code, just being re-factored for the better.

**Comment by oliversalzburg** [comment]
We *are* actually using tasks for read-only transactions. I was just using the term *transaction* broadly.

Let me give an example of what we're dealing with. We have an entity "person". All properties that are relevant to a person are not stored within the same table, they require several in fact.

However, we have a REST API that pretends like all the data stored in all those tables are a single entity. So when we receive a new entity through the API, we need to update all the tables in a transaction. Because we had even more trouble with resource conflicts in the past, we're now setting every transaction as `SERIALIZABLE`, even though it doesn't really matter, as we now avoid all concurrency anyway.

Now, due to the nature of promises and how queries are dispatched, if there are a lot of API requests, it is possible for multiple of these transactions to lock the same resource and transactions get rejected. Retrying transactions is rather complicated and we decided to take the easy way out and simply queue all transactions.

This queuing is already in place at the API level, so it doesn't know if the data layer would only dispatch read-only tasks.

So, yeah, we still have a lot of work to do :P

**Comment by vitaly-t** [comment]
I would be interested to isolate the issue to the PostgeSQL level. I don't think that promises or lower-level drivers that we use are the reason. I am looking at the use of `SERIALIZABLE` and transaction isolation, but it would have to be able to replicate the issue right from pgAdmin UI.

**Comment by vitaly-t** [comment]
@oliversalzburg have you made any progress investigating PostgreSQL transactions? It seems like a very important topic, given that you are running in transaction lock-ups. I really would like to know how to avoid it, and if possible, improve the library to that end. If you do have something to share, please open a new issue for that.

**Comment by vitaly-t** [comment]
@oliversalzburg since you never replied... I have done some of my own research in this area, and added a new feature to the library with 2.5.0 release - [Configurable Transactions](https://github.com/vitaly-t/pg-promise#configurable-transactions).

**Comment by oliversalzburg** [comment]
@vitaly-t Sorry, I wasn't able to look into the topic again and I don't know when I can allocate time to investigate this any further.

We're generally dealing with 2 types of concurrency issues in our codebase:

1. Connection pool exhaustion - Caused by running too many concurrent, read-only queries.

  We solve this with `Promise.map( foo, { concurrency : n } )` or similar approaches.
2. Resource conflicts - Cause by running concurrent transactions that write to the same tables.

  We solve this by only allowing a single request, which can cause these types of transactions, to enter into the data abstraction layer.

The only thing that I could see as beneficial in the library would be a mechanism that can retry transactions if `ERROR:  could not serialize access due to concurrent update` occurs. But I'm not sure if that would be a smart move in production. I prefer to control the behavior at a higher level, where we have full understanding of all the entities we're manipulating, how they affect each other and what measures should be taken to resolve conflicts.

**Comment by vitaly-t** [comment]
@oliversalzburg About the first issue, how large a pool size are you using?

As for the second issue, have you seen this one? - [13.2.2. Repeatable Read Isolation Level](http://www.postgresql.org/docs/9.4/static/transaction-iso.html#XACT-REPEATABLE-READ)

It explains the exact issue you are having and how to overcome it.

And if you want to repeat a transaction in those cases, why not extend the protocol with a corresponding method, something like this - 

```js
var options = {
    extend: function (obj) {
        obj.txAuto = function (p1, p2) {
            function loop() {
                return obj.tx(p1, p2)
                    .then(function (data) {
                        return data;
                    })
                    .catch(function (reason) {
                        if (typeof reason === 'string' && reason.indexOf('could not serialize access due to concurrent update') >= 0) {
                            // must repeat the attempt;
                            return loop();
                        } else {
                            return promise.reject(reason);
                        }
                    });
            }

            return loop();
        };
    }
};
```

**Comment by oliversalzburg** [comment]
It really doesn't matter how large of a pool size we choose. If we don't limit our concurrency, we'll always have actions that will exceed it. Since we changed our approach, we're fine the default pool size, whatever that is.

And, yeah, I know about repeatable transactions. The point is this:
> When an application receives this error message, it should abort the current transaction and retry the whole transaction from the beginning.
We *have* to repeat them ourselves. We don't even want to run into that situation. Who knows how often we have to retry them until they succeed? What do we do with the request that caused the transaction in the meanwhile? The situation must be avoided IMHO.

What would be great if "the system" could identify a transaction that would cause a conflict and queue it. Likewise, it should be able to identify transactions that *won't* cause conflicts and just send them out.

Sending out as many transaction as you want and retrying the ones that failed until they succeed does not seem like a clean approach.

**Comment by vitaly-t** [comment]
>We have to repeat them ourselves. We don't even want to run into that situation. Who knows how often we have to retry them until they succeed? What do we do with the request that caused the transaction in the meanwhile? The situation must be avoided IMHO.

I believe that to solve this is to consider: the request data, the changes needed in the database, the database structure, and transactions in the last. In all, this is an architectural solution, based on good understanding of the system and the requirements. I wouldn't expect a generic solution for this. This is the nitty-gritty of the software architecture.

>What would be great if "the system" could identify a transaction that would cause a conflict and queue it. Likewise, it should be able to identify transactions that won't cause conflicts and just send them out.

However this is approached, it would be way outside of what this library can offer, it is a whole different level of problems. This library can only offer the low-level facility as supported by PostgreSQL. The rest is in your hands.

**Comment by oliversalzburg** [comment]
Heheh, I totally agree :)

**Comment by vitaly-t** [comment]
@oliversalzburg I'm biting the bullet on support for generators, planning to release 2.6 today with the initial generators support ;)

**Comment by oliversalzburg** [comment]
Our whole codebase is still on ES5 :( Thanks though.

**Comment by vitaly-t** [comment]
Are you using NodeJS prior to 4.x?

**Comment by oliversalzburg** [comment]
No, we're using 4.2. But a lot of our code is for the browser and mobile devices. If we want to move to ES6, we have to introduce every developer to it and allow it in all areas of our code base. And then you have potential conflicts with dependencies or development tools…

**Comment by vitaly-t** [comment]
I see. Well, the good news is, generators within pg-promise are, of course, optional, which means you can write code with and without them at the same time. I'm publishing it is within minutes ;)

**Comment by vitaly-t** [comment]
@oliversalzburg the new release with support for ES6 generators is out! :) Check out for details: [Generators](https://github.com/vitaly-t/pg-promise#generators).

**Comment by vitaly-t** [comment]
@oliversalzburg if you have time later, I have published a blog on this: [PostgreSQL with ES6 generators](https://vitalyt.wordpress.com/2015/12/01/postgres-with-es6-generators/)
